File: encrypt_decrypt_doc.py
import os
import sys
import logging
import argparse
import configparser
from datetime import datetime
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding
from docx import Document

def generate_key_pair(private_key_file, public_key_file):
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=2048,
        backend=default_backend()
    )
    public_key = private_key.public_key()

    with open(private_key_file, "wb") as key_file:
        key_file.write(private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.TraditionalOpenSSL,
            encryption_algorithm=serialization.NoEncryption()
        ))

    with open(public_key_file, "wb") as key_file:
        key_file.write(public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        ))

def load_key_from_file(file_name):
    with open(file_name, "rb") as key_file:
        key_data = key_file.read()
        key = serialization.load_pem_private_key(key_data, password=None, backend=default_backend())
    return key

def encrypt_document(document_path, public_key, log_file):
    logging.info(f"Encrypting document: {document_path}")
    with open(document_path, 'rb') as doc_file:
        document_content = doc_file.read()

    encrypted_content = public_key.encrypt(
        document_content,
        padding.OAEP(
            mgf=padding.MGF1(algorithm=hashes.SHA256()),
            algorithm=hashes.SHA256(),
            label=None
        )
    )

    encrypted_file_path = document_path + ".encrypted"
    with open(encrypted_file_path, 'wb') as encrypted_file:
        encrypted_file.write(encrypted_content)

    logging.info(f"Document encrypted: {encrypted_file_path}")
    log_file.write(f"{datetime.now()} - Document encrypted: {encrypted_file_path}\n")

def decrypt_document(encrypted_document_path, private_key, log_file):
    logging.info(f"Decrypting document: {encrypted_document_path}")
    with open(encrypted_document_path, 'rb') as encrypted_file:
        encrypted_content = encrypted_file.read()

    decrypted_content = private_key.decrypt(
        encrypted_content,
        padding.OAEP(
            mgf=padding.MGF1(algorithm=hashes.SHA256()),
            algorithm=hashes.SHA256(),
            label=None
        )
    )

    original_document_path = encrypted_document_path.replace('.encrypted', '')
    with open(original_document_path, 'wb') as decrypted_file:
        decrypted_file.write(decrypted_content)

    logging.info(f"Document decrypted: {original_document_path}")
    log_file.write(f"{datetime.now()} - Document decrypted: {original_document_path}\n")

def process_documents(directory_path, key_file, log_file, action):
    if action == 'encrypt':
        logging.info("Starting encryption process.")
        public_key = load_key_from_file(key_file)
        for filename in os.listdir(directory_path):
            if filename.endswith('.doc') or filename.endswith('.docx'):
                file_path = os.path.join(directory_path, filename)
                encrypt_document(file_path, public_key, log_file)
        logging.info("Encryption process completed.")
    elif action == 'decrypt':
        logging.info("Starting decryption process.")
        private_key = load_key_from_file(key_file)
        for filename in os.listdir(directory_path):
            if filename.endswith('.encrypted'):
                file_path = os.path.join(directory_path, filename)
                decrypt_document(file_path, private_key, log_file)
        logging.info("Decryption process completed.")

def main():
    parser = argparse.ArgumentParser(description='Encrypt or decrypt documents.')
    parser.add_argument('--action', choices=['encrypt', 'decrypt'], required=True, help='Action to perform: encrypt or decrypt.')
    parser.add_argument('--directory', required=True, help='Directory path containing documents.')
    parser.add_argument('--key-file', required=True, help='Key file path (private key for decryption, public key for encryption).')
    parser.add_argument('--log-file', default='encryption.log', help='Log file path.')
    args = parser.parse_args()

    logging.basicConfig(filename=args.log_file, level=logging.INFO)

    process_documents(args.directory, args.key_file, args.log_file, args.action)

if __name__ == "__main__":
    main()

File: encrypt_decrypt_doc.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <filesystem>
#include <openssl/pem.h>
#include <openssl/rsa.h>
#include <openssl/err.h>
#include <openssl/x509.h>
#include <vector>
#include <iomanip>
#include <ctime>

namespace fs = std::filesystem;

void handleErrors() {
    ERR_print_errors_fp(stderr);
    abort();
}

void generateKeyPair(const std::string& privateKeyFile, const std::string& publicKeyFile) {
    RSA* rsa = RSA_generate_key(2048, RSA_F4, nullptr, nullptr);

    BIO* pbio = BIO_new_file(privateKeyFile.c_str(), "w+");
    PEM_write_bio_RSAPrivateKey(pbio, rsa, nullptr, nullptr, 0, nullptr, nullptr);
    BIO_free_all(pbio);

    BIO* pubio = BIO_new_file(publicKeyFile.c_str(), "w+");
    PEM_write_bio_RSAPublicKey(pubio, rsa);
    BIO_free_all(pubio);

    RSA_free(rsa);
}

RSA* loadPrivateKey(const std::string& keyFile) {
    BIO* bio = BIO_new_file(keyFile.c_str(), "rb");
    RSA* rsa = PEM_read_bio_RSAPrivateKey(bio, nullptr, nullptr, nullptr);
    BIO_free(bio);
    return rsa;
}

RSA* loadPublicKey(const std::string& keyFile) {
    BIO* bio = BIO_new_file(keyFile.c_str(), "rb");
    RSA* rsa = PEM_read_bio_RSAPublicKey(bio, nullptr, nullptr, nullptr);
    BIO_free(bio);
    return rsa;
}

std::vector<unsigned char> encrypt(const std::string& data, RSA* publicKey) {
    std::vector<unsigned char> encrypted(RSA_size(publicKey));
    int result = RSA_public_encrypt(data.size(), reinterpret_cast<const unsigned char*>(data.c_str()),
                                    encrypted.data(), publicKey, RSA_PKCS1_OAEP_PADDING);
    if (result == -1) handleErrors();
    return encrypted;
}

std::string decrypt(const std::vector<unsigned char>& encryptedData, RSA* privateKey) {
    std::vector<unsigned char> decrypted(RSA_size(privateKey));
    int result = RSA_private_decrypt(encryptedData.size(), encryptedData.data(),
                                      decrypted.data(), privateKey, RSA_PKCS1_OAEP_PADDING);
    if (result == -1) handleErrors();
    return std::string(reinterpret_cast<char*>(decrypted.data()), result);
}

void encryptDocument(const fs::path& documentPath, RSA* publicKey) {
    std::ifstream docFile(documentPath, std::ios::binary);
    std::ostringstream ss;
    ss << docFile.rdbuf();
    std::string documentContent = ss.str();
    docFile.close();

    std::vector<unsigned char> encryptedContent = encrypt(documentContent, publicKey);
    fs::path encryptedFilePath = documentPath;
    encryptedFilePath += ".encrypted";

    std::ofstream encryptedFile(encryptedFilePath, std::ios::binary);
    encryptedFile.write(reinterpret_cast<char*>(encryptedContent.data()), encryptedContent.size());
    encryptedFile.close();

    std::cout << "Document encrypted: " << encryptedFilePath << std::endl;
}

void decryptDocument(const fs::path& encryptedDocumentPath, RSA* privateKey) {
    std::ifstream encryptedFile(encryptedDocumentPath, std::ios::binary);
    std::vector<unsigned char> encryptedContent((std::istreambuf_iterator<char>(encryptedFile)),
                                               std::istreambuf_iterator<char>());
    encryptedFile.close();

    std::string originalDocumentContent = decrypt(encryptedContent, privateKey);
    fs::path originalDocumentPath = encryptedDocumentPath;
    originalDocumentPath.replace_extension();

    std::ofstream decryptedFile(originalDocumentPath, std::ios::binary);
    decryptedFile.write(originalDocumentContent.c_str(), originalDocumentContent.size());
    decryptedFile.close();

    std::cout << "Document decrypted: " << originalDocumentPath << std::endl;
}

void processDocuments(const fs::path& directoryPath, const std::string& keyFile, const std::string& action) {
    RSA* key = (action == "encrypt") ? loadPublicKey(keyFile) : loadPrivateKey(keyFile);
    if (!key) {
        std::cerr << "Error loading key file." << std::endl;
        return;
    }

    for (const auto& entry : fs::directory_iterator(directoryPath)) {
        if (entry.is_regular_file()) {
            if (action == "encrypt" && (entry.path().extension() == ".doc" || entry.path().extension() == ".docx")) {
                encryptDocument(entry.path(), key);
            } else if (action == "decrypt" && entry.path().extension() == ".encrypted") {
                decryptDocument(entry.path(), key);
            }
        }
    }

    RSA_free(key);
}

int main(int argc, char* argv[]) {
    if (argc < 5) {
        std::cerr << "Usage: " << argv[0] << " <action: encrypt/decrypt> <directory> <key file> <log file>" << std::endl;
        return 1;
    }

    std::string action = argv[1];
    std::string directory = argv[2];
    std::string keyFile = argv[3];
    std::string logFile = argv[4];

    // Log operations to console (or you could implement to write to logFile)
    std::cout << "Starting " << action << " process." << std::endl;
    processDocuments(directory, keyFile, action);
    std::cout << action << " process completed." << std::endl;

    return 0;
}
